### PR TITLE
chore: add alerts state history query service impl

### PR DESCRIPTION
**Pull request by srikanthccv**
### Summary

Part of https://github.com/SigNoz/signoz/issues/4609

PRD for triggered alerts history https://www.notion.so/signoz/Triggered-alerts-history-88271688c980434da1c2b72c44b2b546

The API contract https://www.notion.so/signoz/Alerts-history-API-contract-a43bd02b54084e02ae15bed257d35584

This includes most of the capturing logic when an alert is triggered and nearly finished API.

### CODE REVIEW

**Comment by nityanandagohain** [comment]
Few questions:-

* Did we choose clickhouse as a datstore for this due to the functions like TTL etc, or was there some other reason.
* Also will it be better to keep clickhouse migrations in otel-collector or it's fine here ?

**Comment by srikanthccv** [comment]
1. TTL + doing any analytical stuff is easier on ClickHouse than SQLite
2. I would prefer to keep the migration here itself for now.

There are new set of design changes and will probably need API changes also so I am going to move this to draft till it's ready.

File: pkg/query-service/model/v3/v3.go [hunk 1]
```diff
@@ -1142,3 +1142,92 @@ type MetricMetadataResponse struct {
...
+	Limit   int64      `json:"limit"`
+	Order   string     `json:"order"`
+}
+
+func (r *QueryRuleStateHistory) Validate() error {
+	if r.Start == 0 || r.End == 0 {
+		return fmt.Errorf("start and end are required")
+	}
+	if r.Offset < 0 || r.Limit < 0 {
+		return fmt.Errorf("offset and limit must be greater than 0")
```
Comment (nityanandagohain) [inline pkg/query-service/model/v3/v3.go:1197]: set default limit if not provided?

File: pkg/query-service/app/server.go [hunk 2]
```diff
@@ -147,6 +148,13 @@ func NewServer(serverOptions *ServerOptions) (*Server, error) {
 		return nil, err
 	}
 
+	go func() {
+		err = migrate.ClickHouseMigrate(reader.GetConn(), serverOptions.Cluster)
+		if err != nil {
+			zap.L().Error("error creating history table", zap.Error(err))
```
Comment (nityanandagohain) [inline pkg/query-service/app/server.go:154]: same: error can be error running migration

File: pkg/query-service/rules/promRule.go [hunk 3]
```diff
@@ -470,19 +482,62 @@ func (r *PromRule) Eval(ctx context.Context, ts time.Time, queriers *Queriers) (
...
+				itemsToAdd[idx].OverallState = currentState.String()
+			}
+			itemsToAdd[idx].OverallStateChanged = true
+		}
+	}
+
+	if len(itemsToAdd) > 0 && r.reader != nil {
+		err := r.reader.AddRuleStateHistory(ctx, itemsToAdd)
+		if err != nil {
+			zap.L().Error("error adding to itemsToAdd", zap.Error(err), zap.Any("itemsToAdd", itemsToAdd))
```
Comment (nityanandagohain) [inline pkg/query-service/rules/promRule.go:537]: error adding to to history table ?

File: pkg/query-service/migrate/migate.go [hunk 4]
```diff
@@ -77,3 +80,89 @@ func Migrate(dsn string) error {
...
+    state_changed Bool,
+    unix_milli Int64 CODEC(Delta(8), ZSTD(1)),
+    fingerprint UInt64 CODEC(ZSTD(1)),
+    value Float64 CODEC(Gorilla, ZSTD(1)),
+    labels String CODEC(ZSTD(5)),
+)
+ENGINE = MergeTree
+PARTITION BY toDate(unix_milli / 1000)
+ORDER BY (rule_id, unix_milli)
+TTL toDateTime(unix_milli / 1000) + toIntervalSecond(15552000)
```
Comment (nityanandagohain) [inline pkg/query-service/migrate/migate.go:104]: keep the ttl configurable ? with default 180 days

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 5]
```diff
@@ -5002,6 +5004,319 @@ func (r *ClickHouseReader) LiveTailLogsV3(ctx context.Context, query string, tim
...
 }
 
+func (r *ClickHouseReader) AddRuleStateHistory(ctx context.Context, ruleStateHistory []v3.RuleStateHistory) error {
+	var statement driver.Batch
+	var err error
+
+	statement, err = r.db.PrepareBatch(ctx, fmt.Sprintf("INSERT INTO %s.%s (rule_id, rule_name, overall_state, overall_state_changed, state, state_changed, unix_milli, labels, fingerprint, value) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)",
+		signozHistoryDBName, ruleStateHistoryTableName))
+
+	defer func() {
```
[the remark this answers is not in the document]
Comment (srikanthccv) [reply pkg/query-service/app/clickhouseReader/reader.go:5134]: done

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 6]
```diff
@@ -5002,6 +5004,319 @@ func (r *ClickHouseReader) LiveTailLogsV3(ctx context.Context, query string, tim
...
+		return nil, err
+	}
+
+	return history, nil
+}
+
+func (r *ClickHouseReader) ReadRuleStateHistoryTopContributorsByRuleID(
+	ctx context.Context, ruleID string, params *v3.QueryRuleStateHistory) ([]v3.RuleStateHistoryContributor, error) {
+	// Show the top 10 contributors for a rule
+	query := fmt.Sprintf(`SELECT
```
[the remark this answers is not in the document]
Comment (srikanthccv) [reply pkg/query-service/app/clickhouseReader/reader.go:5233]: I am enforcing that client must send their desired limit.

File: pkg/query-service/model/v3/v3.go (hunk 1 above)
Comment (srikanthccv) [reply pkg/query-service/model/v3/v3.go:1209]: I think the client should consciously make decisions about how much it wants to consume so I am not adding any default.

File: pkg/query-service/migrate/migate.go (hunk 4 above)
Comment (srikanthccv) [reply pkg/query-service/migrate/migate.go]: Not through env or something but I made it configurable. One can always update the default value for `_retention_days` for their desired retention.

File: pkg/query-service/app/clickhouseReader/reader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:5233]: no I am saying there is no limit added to this SQL

[the remark this answers is not in the document]
Comment (srikanthccv) [reply pkg/query-service/app/clickhouseReader/reader.go:5233]: Ah yes, this was supposed to be top 10 initially but later we decided to show everything. The number of entries in the groupby should be limited.

[the remark this answers is not in the document]
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:5233]: okay so we have decided not to add a limit here ?

[the remark this answers is not in the document]
Comment (srikanthccv) [reply pkg/query-service/app/clickhouseReader/reader.go:5233]: yes, we show all sorted by their occurrence

**Review by nityanandagohain — APPROVED** [review]

**Pull request merged**